### PR TITLE
feat: report different request errors

### LINKED ISSUE #3855: report transaction response error accurately
### Describe the feature

all variants are currently reported the same:

https://github.com/paradigmxyz/reth/blob/7476120d8170d7269ec8525f173edac995e12385/crates/interfaces/src/p2p/error.rs#L76-L85

https://github.com/paradigmxyz/reth/blob/7476120d8170d7269ec8525f173edac995e12385/crates/net/network/src/transactions.rs#L540-L542

even though we have different weights for timeout and bad message

## TODO

report peer based on error variant

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by altugbakan**
Fixes #3855 

This pull request handles different transaction response errors.

### CODE REVIEW

File: crates/net/network/src/transactions.rs [hunk 1]
```diff
@@ -472,10 +475,27 @@ where
         }
     }
 
-    fn report_bad_message(&self, peer_id: PeerId) {
-        trace!(target: "net::tx", ?peer_id, "Penalizing peer for bad transaction");
+    fn report_bad_message(&self, peer_id: PeerId, req_err: Option<RequestError>) {
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs:478]: we can keep this and rename it to `on_request_error` and we need another function that accepts a `ReputationChangeKind`.
then this function here does not need an option anymore

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/transactions.rs (hunk 1 above)
Comment (mattsse) [reply crates/net/network/src/transactions.rs:478]: > Hey, just to make it clear, should I revert the changes on the report_bad_message

something in between, revert most changes for this but rename to `report_message(kind)` and add a `ReputationChangeKind` the new `on_request_error ` function then calls this

Comment (altugbakan) [reply crates/net/network/src/transactions.rs]: I did some changes regarding your comment, did I get it right?

File: crates/net/network/src/transactions.rs [hunk 2]
```diff
@@ -475,29 +475,37 @@ where
...
+        let msg = match kind {
+            ReputationChangeKind::BadProtocol => "Penalizing peer for bad protocol rule",
+            ReputationChangeKind::Timeout => "Penalizing peer for timeout",
+            ReputationChangeKind::Dropped => "Penalizing peer for dropped connection",
+            ReputationChangeKind::BadTransactions => "Penalizing peer for bad transaction",
+            ReputationChangeKind::AlreadySeenTransaction => {
+                "Penalizing peer for already seen transaction"
             }
+            ReputationChangeKind::BadBlock => "Penalizing peer for bad block",
+            ReputationChangeKind::BadMessage => "Penalizing peer for bad message",
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs]: I think using the debug impl for this is sufficient, so no need to derive a message

File: crates/net/network/src/transactions.rs [hunk 3]
```diff
@@ -475,29 +475,37 @@ where
...
-                "Penalizing peer for timeout"
-            }
-            Some(RequestError::ChannelClosed | RequestError::ConnectionDropped) => {
-                self.network.reputation_change(peer_id, ReputationChangeKind::Dropped);
-                "Penalizing peer for dropped connection"
-            }
-            Some(RequestError::BadResponse) | None => {
-                self.network.reputation_change(peer_id, ReputationChangeKind::BadTransactions);
-                "Penalizing peer for bad transaction"
+    fn report_message(&self, peer_id: PeerId, kind: ReputationChangeKind) {
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs:478]:
```suggestion
    fn report_peer(&self, peer_id: PeerId, kind: ReputationChangeKind) {
```

File: crates/net/network/src/transactions.rs [hunk 4]
```diff
@@ -475,29 +475,37 @@ where
...
+    fn on_request_error(&self, peer_id: PeerId, req_err: RequestError) {
+        let kind = match req_err {
+            RequestError::UnsupportedCapability => ReputationChangeKind::BadProtocol,
+            RequestError::Timeout => ReputationChangeKind::Timeout,
+            RequestError::ChannelClosed | RequestError::ConnectionDropped => {
+                ReputationChangeKind::Dropped
+            }
+            RequestError::BadResponse => ReputationChangeKind::BadTransactions,
+        };
+        self.network.reputation_change(peer_id, kind);
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs:505]: this needs to go to the other function

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits

File: crates/net/network/src/transactions.rs [hunk 5]
```diff
@@ -492,7 +507,7 @@ where
     fn on_bad_import(&mut self, hash: TxHash) {
         if let Some(peers) = self.transactions_by_peers.remove(&hash) {
             for peer_id in peers {
-                self.report_bad_message(peer_id);
+                self.on_request_error(peer_id, RequestError::BadResponse);
```
Comment (mattsse) [inline crates/net/network/src/transactions.rs:510]: last nit: we want to call the reputation kind function here, with `BadTransactions`

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**